File: stage3_tmean.py
#!/usr/bin/env python3
#
# Generate mean temperature, taken as the average of tmax and tmin. Low-memory
# version.
#
# Contact: Edgar Castro <edgar_castro@g.harvard.edu>

import csv
import gzip
import os

import tqdm

DEFAULT_TMAX_FILENAME = "mean_tmax.csv.gz"
DEFAULT_TMIN_FILENAME = "mean_tmin.csv.gz"
DEFAULT_TMEAN_FILENAME = "mean_tmean.csv.gz"


def generate_tmean(tmin_path: str, tmax_path: str, output_path: str):
    temp_path = "{}.part".format(output_path)

    with gzip.open(tmin_path, "rt") as tmin_fp, \
            gzip.open(tmax_path, "rt") as tmax_fp, \
            gzip.open(temp_path, "wt") as output_fp:
        tmax_reader = csv.DictReader(tmax_fp)
        tmin_reader = csv.DictReader(tmin_fp)

        id_column = tmax_reader.fieldnames[0]

        tmean_writer = csv.DictWriter(output_fp, fieldnames=[id_column, "date", "tmean"])
        tmean_writer.writeheader()

        for tmax_row in tqdm.tqdm(tmax_reader):
            tmin_row = next(tmin_reader)

            if not tmax_row:
                break

            # assert tmax_row[id_column] == tmin_row[id_column]
            # assert tmax_row["date"] == tmin_row["date"]

            tmean_writer.writerow({
                id_column: tmax_row[id_column],
                "date": tmax_row["date"],
                "tmean": (float(tmax_row["value"]) + float(tmin_row["value"])) / 2
            })

    os.rename(temp_path, output_path)


if __name__ == "__main__":
    import argparse
    import glob

    parser = argparse.ArgumentParser()
    parser.add_argument("-t", "--tmin-file", default=None)
    parser.add_argument("-T", "--tmax-file", default=None)
    parser.add_argument("-o", "--output", default=None)
    args = parser.parse_args()

    # Input given: convert the given file
    if all([args.tmin_file, args.tmax_file, args.output]):
        generate_tmean(args.tmin_file, args.tmax_file, args.output)

    # No input given: determine what files need to be converted by looking for
    # missing files in the expected daymet-aggregation output directory
    # hierarchy
    else:
        for aggregated_directory in glob.glob("output/aggregated-combined/*"):
            extra_directory = aggregated_directory.replace("aggregated-combined", "extra")

            tmax_path = os.path.join(aggregated_directory, DEFAULT_TMAX_FILENAME)
            tmin_path = os.path.join(aggregated_directory, DEFAULT_TMIN_FILENAME)
            tmean_path = os.path.join(extra_directory, DEFAULT_TMEAN_FILENAME)

            for path in [tmax_path, tmin_path]:
                if not os.path.isfile(path):
                    raise Exception("ERROR: {} does not exist".format(path))

            if not os.path.isdir(extra_directory):
                os.makedirs(extra_directory)

            if os.path.isfile(tmean_path):
                print("Skipping {}".format(tmean_path))
            else:
                print("Generating {}".format(tmean_path))
                generate_tmean(tmin_path, tmax_path, tmean_path)

File: stage3_tmean.jl
#!/usr/bin/env julia
#
# Generate mean temperature, taken as the average of tmax and tmin. Low-memory
# and fast-running version.
#
# Contact: Edgar Castro <edgar_castro@g.harvard.edu>

using ArgParse
using CodecZlib
using Printf
using ProgressMeter

DEFAULT_TMAX_FILENAME = "mean_tmax.csv.gz"
DEFAULT_TMIN_FILENAME = "mean_tmin.csv.gz"
DEFAULT_TMEAN_FILENAME = "mean_tmean.csv.gz"

function generate_tmean(tmin_path::String, tmax_path::String, output_path::String)
    temp_path = "$output_path.part"
    
    tmin_stream = GzipDecompressorStream(open(tmin_path, "r"))
    tmax_stream = GzipDecompressorStream(open(tmax_path, "r"))
    tmean_stream = GzipCompressorStream(open(temp_path, "w"))

    id_column = split(readline(tmin_stream), ",")[1]
    readline(tmax_stream)

    progress = ProgressUnknown("Generating $output_path:")
    write(tmean_stream, "$id_column,date,tmean\n")
    for tmin_line in eachline(tmin_stream)
        tmin_split = split(tmin_line, ",")
        tmax_split = split(readline(tmax_stream), ",")
        tmean = (parse(Float32, tmin_split[3]) + parse(Float32, tmax_split[3])) / 2
        write(tmean_stream, "$(tmin_split[1]),$(tmin_split[2]),$tmean\n")
        ProgressMeter.next!(progress)
    end
    ProgressMeter.finish!(progress)
    
    mv(temp_path, output_path)
end

arg_settings = ArgParseSettings()
@add_arg_table arg_settings begin
    "--tmin-file", "-t"
    "--tmax-file", "-T"
    "--output", "-o"
end

args = parse_args(arg_settings)

if all(values(args) .!= nothing)
    generate_tmean(args["tmin-file"], args["tmax-file"], args["output"])
else
    for geography_name in readdir("output/aggregated-combined/")
        aggregated_directory = "output/aggregated-combined/$geography_name"       
        extra_directory = "output/aggregated-combined/$geography_name"
        println(aggregated_directory)
    
        tmax_path = "$aggregated_directory/$DEFAULT_TMAX_FILENAME"
        tmin_path = "$aggregated_directory/$DEFAULT_TMIN_FILENAME"
        tmean_path = "$extra_directory/$DEFAULT_TMEAN_FILENAME"
        
        for file in [tmax_path, tmin_path]
            if ! isfile(file)
                throw(ErrorException("File $file does not exist"))
            end
        end
        
        if isfile(tmean_path)
            println("Skipping $tmean_path")
        else
            generate_tmean(tmin_path, tmax_path, tmean_path)
        end
    end
end
